packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts

```diff
@@ -0,0 +1,265 @@
+} from '../repositories/AnomaliesRepository'
+import {
+  LivenessRecordWithSubtype,
+  LivenessRepository,
+} from '../repositories/LivenessRepository'
+import { Interval, calculateIntervals } from '../utils/calculateIntervals'
+import { getProjectsToSync } from '../utils/getProjectsToSync'
+import { groupByType } from '../utils/groupByType'
+
+export interface AnomaliesIndexerIndexerDeps
+  extends Omit<ManagedChildIndexerOptions, 'name'> {
+  livenessRepository: LivenessRepository
+  anomaliesRepository: AnomaliesRepository
+  projects: Project[]
+}
+
+export class AnomaliesIndexer extends ManagedChildIndexer {
+  private readonly SYNC_RANGE = 30
+  constructor(private readonly $: AnomaliesIndexerIndexerDeps) {
+    super({ ...$, name: 'anomalies' })
+  }
+
+  override async update(from: number, to: number): Promise<number> {
+    // we only need to go one day back
+    const maxDepth = UnixTime.now().add(-1, 'days').toStartOf('day').toNumber()
+    if (to <= maxDepth) {
+      this.logger.info('Skipping update', { from, to })
+      return to
+    }
+
+    // only on first run when from equals minTimestamp
+    if (from < maxDepth) {
+      from = maxDepth
+      this.logger.info('Adjusting update range', { from, to })
+    }
+
+    // limit time range to one day if greater
+    const { to: unixTo } = clampRangeToDay(from, to)
+
+    this.logger.info('Calculating anomalies', { unixTo })
+
+    const anomalies = await this.getAnomalies(unixTo)
+
+    await this.$.anomaliesRepository.addOrUpdateMany(anomalies)
+
+    return unixTo.toNumber()
+  }
+
+  override async invalidate(targetHeight: number): Promise<number> {
+    // no need to remove data
+    // safeHeight will be updated to this value
+    return await Promise.resolve(targetHeight)
+  }
+
+  async getAnomalies(to: UnixTime) {
+    const anomalies: AnomaliesRecord[] = []
+
+    const configurations =
+      await this.$.indexerService.getSavedConfigurations<TrackedTxConfigEntry>(
+        'tracked_txs_indexer',
+      )
+
+    const projectsToSync = getProjectsToSync(this.$.projects, configurations)
+
+    // we need data from 2 * SYNC_RANGE past days to calcualate standard deviation
+    const deviationRange = to.add(-1 * this.SYNC_RANGE * 2, 'days')
+
+    for (const project of projectsToSync) {
+      const livenessRecords =
+        await this.$.livenessRepository.getWithSubtypeByProjectIdsWithinTimeRange(
+          project.projectId,
+          deviationRange,
+          to,
+        )
+
+      if (livenessRecords.length === 0) {
+        this.logger.debug('No records found for project', {
+          projectId: project.projectId,
+        })
+        continue
+      }
+
+      this.logger.debug('Liveness records loaded', {
+        projectId: project.projectId,
+        count: livenessRecords.length,
+      })
+
+      const [batchSubmissions, stateUpdates, proofSubmissions] =
+        groupByType(livenessRecords)
+
+      const test = this.detectAnomalies(
```

Comment:
I guess it should be named differently or just used inline inside a fn

Reply:
argh... leftover :)